src/edu/washington/escience/myria/operator/SamplingDistribution.java

```diff
@@ -0,0 +1,320 @@
+      Type col0Type = tb.getSchema().getColumnType(0);
+      Type col1Type = tb.getSchema().getColumnType(1);
+      boolean hasStreamSize = false;
+      Type col2Type = null;
+      if (tb.getSchema().numColumns() > 2) {
+        hasStreamSize = true;
+        col2Type = tb.getSchema().getColumnType(2);
+      }
+      for (int i = 0; i < tb.numTuples(); i++) {
+        int workerID;
+        if (col0Type == Type.INT_TYPE) {
+          workerID = tb.getInt(0, i);
+        } else if (col0Type == Type.LONG_TYPE) {
+          workerID = (int) tb.getLong(0, i);
+        } else {
+          throw new DbException("WorkerID must be of type INT or LONG");
+        }
+        Preconditions.checkState(workerID > 0, "WorkerID must be > 0");
+        // Ensure the future .set(workerID, -) calls will work.
+        for (int j = tupleCounts.size(); j < workerID; j++) {
+          tupleCounts.add(0);
+          streamCounts.add(0);
+        }
+
+        int partitionSize;
+        if (col1Type == Type.INT_TYPE) {
+          partitionSize = tb.getInt(1, i);
+        } else if (col1Type == Type.LONG_TYPE) {
+          partitionSize = (int) tb.getLong(1, i);
+        } else {
+          throw new DbException("PartitionSize must be of type INT or LONG");
+        }
+        Preconditions.checkState(partitionSize >= 0,
+            "Worker cannot have a negative PartitionSize: %s", partitionSize);
+        tupleCounts.set(workerID - 1, partitionSize);
+        totalTupleCount += partitionSize;
+        int streamSize = partitionSize;
+        if (hasStreamSize) {
+          if (col2Type == Type.INT_TYPE) {
+            streamSize = tb.getInt(2, i);
+          } else if (col2Type == Type.LONG_TYPE) {
+            streamSize = (int) tb.getLong(2, i);
+          } else {
+            throw new DbException("StreamSize must be of type INT or LONG");
+          }
+          Preconditions.checkState(partitionSize >= 0,
+              "Worker cannot have a negative StreamSize: %d", streamSize);
+        }
+        streamCounts.set(workerID - 1, streamSize);
+      }
+    }
+    // Convert samplePct to sampleSize if using a percentage sample.
+    if (isPercentageSample) {
+      sampleSize = Math.round(totalTupleCount * (samplePercentage / 100));
+    }
+    Preconditions.checkState(sampleSize >= 0 && sampleSize <= totalTupleCount,
+        "Cannot extract %s samples from a population of size %s", sampleSize,
+        totalTupleCount);
+
+    // Generate a random distribution across the workers.
+    int[] sampleCounts;
+    if (sampleType == SamplingType.WR) {
+      sampleCounts = withReplacementDistribution(tupleCounts, sampleSize);
+    } else if (sampleType == SamplingType.WoR) {
+      sampleCounts = withoutReplacementDistribution(tupleCounts, sampleSize);
+    } else {
+      throw new DbException("Invalid sampleType: " + sampleType);
+    }
+
+    // Build and return a TupleBatch with the distribution.
+    IntColumnBuilder wIdCol = new IntColumnBuilder();
```

Comment:
I think you could save a bit of code by using `TupleBatchBuffer` and `popAny()`.